Redesign EPP Metrics Pipeline to be Model Server Agnostic

BenjaminBraunDev opened the pull request:
This is a redesign how EPP extracts metrics from its model server endpoints in a way that works with any model server that supports prometheus metrics. 

Additionally, so long as the model server implements the [OpenAI API protocol](https://docs.vllm.ai/en/latest/getting_started/quickstart.html#openai-compatible-server) EPP can send inference requests and load balance across InferenceModel objects regardless of the underlying model server architecture.

### What problem is being solved?

The difficult part of enabling EPP to collect metrics for any model server is the translation of metrics from what the model server exposes into the set of metrics we load balance on. While many model serves export their metrics in prometheus format, there is no consistent naming scheme or interface to follow.

The goal of this redesign is to provide a set of assignable modelserver-side prometheus metrics and automatically map them to the `datastore.Metrics` type fields that EPP uses to load balance.

### Why the complexity?

In some cases, a model server's prometheus metric might directly correspond to one of the load-balancing metrics, but other times we need multiple prometheus metrics to derive a single load-balancing metric. Therefore, some form of mapping from the set of common prometheus metrics a server might provide to the load-balancing metrics we need is necessary.

To address this, we provide EPP with several potential runtime flags that define what metrics it has to work with for the model server endpoints it's serving. From there, we check if the flags provided are adequate to derive all the required load balancing metrics. For example:

In order for EPP to derive `KVCacheUsagePercent`, a model server must provide either:
1. `KVCacheUsagePercent` (vLLM provides this directly)
OR
2. `UsedKVCacheBlocks` and `MaxKVCacheBlocks` (Triton has no percentage metric, so provides both of these instead)

Each of these is provided as the latest value of some prometheus metric family, optionally filtered on a set of key/value labels. Going forward, to incorporate a new model server it can either provide a family/labels for the direct metric or alternatively a family/labels for each of the constituent metrics, and EPP will derive the metric from there.

## Breakdown of Changes

### `backend/metrics.go` and new `backend/metrics_spec.go`

`backend` now has a single metrics.go for all model servers, and has an a new field in its `PodMetricsClient` implementation called `MetricMapping`

```
type PodMetricsClientImpl struct {
	MetricMapping *MetricMapping
}
```

This MetricMapping is a set of predefined MetricSpec objects that represent what constituent metrics a model server might provide, the idea is any model server should be able to provide most of these, but perhaps not all, and we should be robust to that without requiring any input from the model server side.

```
// MetricSpec represents a single metric's specification.
type MetricSpec struct {
	MetricName string // Prometheus metric family name
	Labels     map[string]string // Label name -> Label value
}

// MetricMapping holds named MetricSpecs.
type MetricMapping struct {
	AllRequests       *MetricSpec // Option 1
	WaitingRequests   *MetricSpec // Option 2
	RunningRequests   *MetricSpec // Required
	UsedKVCacheBlocks *MetricSpec // Option 1 (part of a group)
	MaxKVCacheBlocks  *MetricSpec // Option 1 (part of a group)
	KVCacheUsage      *MetricSpec // Option 2 (alternative to the group above)
	// LoRA Metrics (vLLM Specific, optional)
	LoraRequestInfo *MetricSpec
}
```

From there, the logic for knowing whether we have substantial prometheus metrics to load balance, and how to derive those load balancing metrics, is in `metrics.go`. To clarify, the goal is not to process prometheus metrics based on the specific model server, but rather abstract away what model server we are scraping and scrape purely based on the subset of prometheus metrics we are given in the EPP runtime flags. This allows us to avoid writing a new metrics.go (and compiling a separate EPP image) for each model sever we want to support, only requiring code changes if a new model server isn't able to provide metrics that "span" the load balancing metrics.

### `main.go`

Here we add flags for the prometheus family/labels. **The format of these flags is modeled after prometheus strings themselves**, that being a metric family follows by a list of key-value pair labels in curly braces:

```
-metricFlagName="metric_family_name{label1=value1,label2=value2}"
```

From there, `metrics.go` will scrape the latest metrics in that family that has at least all the specified labels. For example, for triton, the EPP flags look like this:

from args section of `config/manifests/triton/ext_proc.yaml`
```
- -allRequestsMetric
- "nv_trt_llm_request_metrics{request_type=active}"
- -runningRequestsMetric
- "nv_trt_llm_request_metrics{request_type=scheduled}"
- -usedKVCacheBlocksMetric
- "nv_trt_llm_kv_cache_block_metrics{kv_cache_block_type=used}"
- -maxKVCacheBlocksMetric
- "nv_trt_llm_kv_cache_block_metrics{kv_cache_block_type=max}"
```

Whereas vLLM has separate metric families for each of its metrics (except lora metrics, which are handled differently) and the flags look like this:

from args section of `config/manifests/vllm/ext_proc.yaml`
```
- -waitingRequestsMetric
- "vllm:num_requests_waiting"
- -runningRequestsMetric
- "vllm:num_requests_running"
- -kVCacheUsageMetric
- "vllm:gpu_cache_usage_perc"
- -loraRequestInfoMetric
- "vllm:lora_requests_info"
```

In either case, we can deploy EPP for either Triton or vLLM using the same image, only changing the flags.

## Limitations

Currently, this still requires EPP to be re-run to switch model server is being inferenced, the next logical step is to allow gateway to choose between EPP services based on request, or perhaps allow a single running instance of EPP to support multiple model server types.

Additionally, if a model server has new prometheus metrics that can still be derived into our load balancing metrics in some way, we would have to add that logic in `metrics.go`.

LoRA metrics are also handled differently and are treated as a special case for vLLM. There is a single metric family, and for each entry the value is a time stamp, with the actual metric information being in the labels that entry has. If a model server happens to use this same system as vLLM, is can provide a metric family name for the `-loraRequestInfoMetric` flag, and it will be handled in the same way vLLM is, however it's unclear if this is a standard approach for LoRA metrics.

hzxuzhonghu commented on pkg/epp/backend/metrics.go:
```diff
@@ -0,0 +1,321 @@
+	logger := log.FromContext(ctx)
+	loggerDefault := logger.V(logutil.DEFAULT)
+
+	url := "http://" + existing.Address + ":" + strconv.Itoa(int(port)) + "/metrics"
```
Not related to this change, i am curious  if this is not always the same url for different inference engine

ahg-g replied:
We need to make this part of the model server protocol @liu-cong

BenjaminBraunDev replied:
resolving, out of scope for this PR

hzxuzhonghu reviewed (commented):
I would think it is better to define a MetricsFetcher interface, for each inference framework, there should be a implement. 

By passing the metric name from args looks like hardcoding, it can be difficult if we want to support multi inference engine simultaneously later

BenjaminBraunDev commented:
@liu-cong @ahg-g 

Removed intermediate metrics and triton support (for now, until we make the changes in trtllm backend). Made the metrics mapping 1-to-1 with the metrics we support for load balancing, and made defaults the metric families for vLLM.

Also, we should update the [Quickstart Guide](https://gateway-api-inference-extension.sigs.k8s.io/guides/) to account to account for the movement of ext_proc into the vllm directory here, and also because it's out of date with some of the file paths. For instance, the [vLLM deployment](https://gateway-api-inference-extension.sigs.k8s.io/guides/#deploy-sample-model-server) was broken into 2 separate yamls, so this command doesn't work anymore:

```
https://gateway-api-inference-extension.sigs.k8s.io/guides/#deploy-sample-model-server
```

We should change that, and [here](https://gateway-api-inference-extension.sigs.k8s.io/guides/#deploy-the-inference-extension-and-inferencepool) change (I moved this here since currently a new ext_proc.yaml will need to be made for each model server, setting the appropriate metric flags. I think we should keep vLLM specific things inside a vllm directory.)

```
kubectl apply -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/manifests/ext_proc.yaml
```
to
```
kubectl apply -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/manifests/vllm/ext_proc.yaml
```

liu-cong commented on pkg/epp/backend/metrics_spec.go:
```diff
@@ -0,0 +1,117 @@
+// MetricSpec represents a single metric's specification.
+type MetricSpec struct {
+	MetricName string
+	Labels     map[string]string // Label name -> Label value
```
do we want to have a map or just a label key and value tuple?

BenjaminBraunDev replied:
I used a map as it's the most accurate data structure to describe the labels section of prometheus metrics, the labels are unique and the keys are all strings.

BenjaminBraunDev replied:
it's also quick to lookup label values with a map

liu-cong commented on pkg/epp/backend/metrics_spec.go:
```diff
@@ -0,0 +1,117 @@
+//	"metric_name"
+//	"metric_name{label1=value1}"
+//	"metric_name{label1=value1,label2=value2}"
+func stringToMetricSpec(specStr string) (*MetricSpec, error) {
```
I suggest just adding the metricLabelKey and metricLabelValue as flags, so we don't need to do this heavy parsing. It's not pretty (each metric needs 3 flags: name, label key and value), but it's clear and less error prone.

BenjaminBraunDev replied:
In some cases there may be multiple key value pairs, like with the lora metrics, we read off several labels all in the same metric.

liu-cong replied:
I don't think we have metrics that need multiple labels to filter, but I see your point of being future proof.

I don't like we are doing complex string parsing, in the meantime I cannot think of an approach that's strictly better. Given you have good coverage of the edge cases, I have confidence on the parsing logic. 

@ahg-g Thoughts?

ahg-g replied:
I am guessing this code is not really getting executed for vllm since we are not selecting specific labels (the labels we select for lora are hardcoded somewhere else). If we have good test coverage it is fine, but again, I would rather have a clear and well defined spec in the model server protocol rather than this.

BenjaminBraunDev replied:
For now, I'm adding the direct metrics in TRT-LLM backend for integration with Triton, the names of these will be added via the epp runtime flags. Once we have that working integration working e2e we can discuss changes to the metrics protocol for gateway API to enforce a specific prometheus metric naming schema for all model servers. Resolving for the scope of this PR.

liu-cong commented on pkg/epp/backend/vllm/metrics.go:
```diff
@@ -34,25 +33,16 @@ import (
-// https://github.com/kubernetes-sigs/gateway-api-inference-extension/tree/main/docs/proposals/003-model-server-protocol
 const (
-	LoraRequestInfoMetricName                = "vllm:lora_requests_info"
+	// Hardcoded vLLM specific LoRA metrics
```
This is not vllm specific, they are defined in the protocol. You can refer to the protocol in the comment: https://github.com/kubernetes-sigs/gateway-api-inference-extension/tree/main/docs/proposals/003-model-server-protocol

BenjaminBraunDev replied:
fixed comment

liu-cong commented on pkg/epp/backend/vllm/metrics.go:
```diff
@@ -232,6 +234,54 @@ func getLatestMetric(logger logr.Logger, metricFamilies map[string]*dto.MetricFa
 }
+
+// getLabeledMetric gets the latest metric with matching labels.
+func getLabeledMetric(logger logr.Logger, mf *dto.MetricFamily, spec MetricSpec) (*dto.Metric, error) {
```
Can we combine this with getLatestMetric, so if the labels are empty, then this does exactly what `getLatestMetric` does

BenjaminBraunDev replied:
getMetrics is essentially that, if there are labels it calls `getLabeledMetric` otherwise `getLatestMetric`

BenjaminBraunDev replied:
done

liu-cong commented on pkg/epp/backend/vllm/metrics.go:
```diff
@@ -63,8 +53,6 @@ func (p *PodMetricsClientImpl) FetchMetrics(
 
-	// Currently the metrics endpoint is hard-coded, which works with vLLM.
-	// TODO(https://github.com/kubernetes-sigs/gateway-api-inference-extension/issues/16): Consume this from InferencePool config.
 	url := "http://" + existing.Address + ":" + strconv.Itoa(int(port)) + "/metrics"
```
Pls keep the TODO. We should make the metrics port and path configurable, but can be a followup.

BenjaminBraunDev commented on cmd/epp/main.go:
```diff
@@ -92,6 +91,17 @@ var (
+	totalQueuedRequestMetric = flag.String("totalQueuedRequestMetric",
+		"vllm:num_requests_waiting",
+		"Prometheus metric for the number of queued requests.")
+	kVCacheUsageMetric = flag.String("kVCacheUsageMetric",
```
done

liu-cong commented on pkg/epp/backend/vllm/metrics.go:
```diff
@@ -34,25 +33,16 @@ import (
 	logutil "sigs.k8s.io/gateway-api-inference-extension/pkg/epp/util/logging"
 )
 
-// Metric names used in the vLLM metrics implementation.
```
pks keep the comment

BenjaminBraunDev replied:
All the variables this comment was referring to have been removed, there's nothing vllm specific anymore. (The LoRA variables correspond to vLLMs LoRA metrics, but these are part of the [LoRA metric protocol](https://github.com/kubernetes-sigs/gateway-api-inference-extension/tree/main/docs/proposals/003-model-server-protocol) we want other servers to implement)

liu-cong commented on pkg/epp/backend/vllm/metrics.go:
```diff
@@ -14,7 +14,6 @@ See the License for the specific language governing permissions and
 */
 
-// Package vllm provides vllm specific pod metrics implementation.
 package vllm
```
let's rename this to metrics

BenjaminBraunDev replied:
Will do, these files were originally named metrics, moved them back to make them diff with the original metrics.go file. There's nothing vllm specific in these files anymore so there's no need to put them in a vllm directory, we should move them out in a follow up PR.

liu-cong commented on pkg/epp/backend/vllm/metrics_spec_test.go:
```diff
@@ -0,0 +1,172 @@
+				return
+			}
+			if tt.want != nil && got != nil { // compare maps directly
+				if tt.want.Labels == nil {
```
Specify an empty map in the test case, don't override

BenjaminBraunDev replied:
done

liu-cong commented on pkg/epp/backend/vllm/metrics_spec.go:
```diff
@@ -0,0 +1,117 @@
+//	"metric_name{label1=value1,label2=value2}"
+func stringToMetricSpec(specStr string) (*MetricSpec, error) {
+	if specStr == "" {
+		return nil, nil // Allow empty strings to represent nil MetricSpecs
```
why do we want a nil spec?

BenjaminBraunDev replied:
A nil spec was for when we didn't have a certain metric, since we could support some metrics not being specified if we could make them from other metrics we did have (e.g. either kvCachePercent or both maxBlocks and usedBlocks). This is no longer necessary since we will have a 1-to-1 mapping and all metrics must be specified. Will remove this.

liu-cong commented on pkg/epp/backend/vllm/metrics_spec_test.go:
```diff
@@ -0,0 +1,172 @@
+				t.Errorf("stringToMetricSpec() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if tt.want != nil && got != nil { // compare maps directly
```
I would expect the function never return a nil spec without an error, so if you update that, then the tests can be simplified.

BenjaminBraunDev replied:
Yes, as of our sync Friday we will have a 1-to-1 map and every metric must be specified so no more nil specs. Will remove these cases.

liu-cong commented on pkg/epp/backend/vllm/metrics_spec_test.go:
```diff
@@ -0,0 +1,172 @@
+				if tt.want.Labels == nil {
+					tt.want.Labels = make(map[string]string)
+				}
+				if !reflect.DeepEqual(got.MetricName, tt.want.MetricName) {
```
This is just comparing strings, so you can just use ==

liu-cong commented on pkg/epp/backend/vllm/metrics_test.go:
```diff
@@ -17,234 +17,573 @@ limitations under the License.
+
+// TestFetchMetrics is a basic integration test.  A more complete test would mock
+// the HTTP client.
+func TestFetchMetrics(t *testing.T) {
```
This test case is justing test an connection error message? I don't we should add this...

liu-cong commented on pkg/epp/backend/vllm/metrics_test.go:
```diff
@@ -17,234 +17,573 @@ limitations under the License.
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.shouldPanic {
```
why would the test cause panic?

BenjaminBraunDev replied:
I think this is a remnant of a removed case, will remove this from the test vars

liu-cong commented on pkg/epp/backend/vllm/metrics_test.go:
```diff
@@ -17,234 +17,573 @@ limitations under the License.
+				}
+			} else {
+				if err != nil {
+					t.Errorf("getMetric() unexpected error: %v", err)
```
You probbaly should do Fatal here, as the gotMetric can be nil if there is an error

BenjaminBraunDev replied:
done

liu-cong commented on pkg/epp/backend/vllm/metrics_test.go:
```diff
@@ -17,234 +17,573 @@ limitations under the License.
+	tests := []struct {
+		name        string
+		spec        MetricSpec
+		wantValue   float64
```
```suggestion
		wantGaugeValue   float64
```

liu-cong commented on pkg/epp/backend/vllm/metrics_test.go:
```diff
@@ -17,234 +17,573 @@ limitations under the License.
 	logutil "sigs.k8s.io/gateway-api-inference-extension/pkg/epp/util/logging"
 )
 
-func TestPromToPodMetrics(t *testing.T) {
```
The existing test cases look good to me, why are we spitting it to TestGetLatestLoraMetric and TestPromToPodMetrics? Is it to increase test coverage? What unique cases are being covered in TestGetLatestLoraMetric?

BenjaminBraunDev replied:
LoRA metrics are scraped differently than other metrics, and this TestGetLatestLoraMetric is specifically a unit test of the function that scrapes LoRA metrics `getLatestLoraMetric` it's essentially so each complex function has a test of its functionality, TestPromToPodMetrics tests the whole function, but mostly has tests for the non-lora metrics. getLatestLoraMetric tests lora metric edge cases.

liu-cong commented on pkg/epp/backend/vllm/metrics_test.go:
```diff
@@ -17,234 +17,573 @@ limitations under the License.
+					t.Errorf("promToPodMetrics() unexpected error: %v", err)
+				}
 			} else {
-				assert.NoError(t, err)
```
The existing use of the assert package is cleaner, let's not change it?

BenjaminBraunDev replied:
reverted

ahg-g commented on pkg/epp/backend/metrics.go:
```diff
@@ -0,0 +1,287 @@
+func (p *PodMetricsClientImpl) getMetric(logger logr.Logger, metricFamilies map[string]*dto.MetricFamily, spec MetricSpec) (*dto.Metric, error) {
+	mf, ok := metricFamilies[spec.MetricName]
+	if !ok {
+		logger.V(logutil.DEFAULT).Error(nil, "Metric family not found", "name", spec.MetricName)
```
avoid logging in any of the functions that are called during metrics scraping, this is done very frequently and if something is off, it will spam the logs

BenjaminBraunDev replied:
done

ahg-g commented on pkg/epp/backend/metrics.go:
```diff
@@ -0,0 +1,287 @@
+		return nil, fmt.Errorf("no metrics found for %q", mf.GetName())
+	}
+
+	logger.V(logutil.TRACE).Info("Latest metric value selected", "value", latest, "metric", mf.GetName())
```
ditto

liu-cong replied:
having a TRACE level log is fine. Agree the library code shouldn't have any other lower level logs than TRACE.

BenjaminBraunDev replied:
replaced non-error DEFAULT logging with TRACE

ahg-g commented on pkg/epp/backend/metrics.go:
```diff
@@ -0,0 +1,287 @@
+	existing *datastore.PodMetrics,
+) (*datastore.PodMetrics, error) {
+	var errs error
+	updated := existing.Clone()
```
@liu-cong perhaps we can address this in your PR, but why are we cloning here? Just create a Metrics (not PodMetrics) struct and return it.

liu-cong replied:
yeah this is addressed in https://github.com/kubernetes-sigs/gateway-api-inference-extension/pull/460/files

BenjaminBraunDev commented:
@ahg-g
> I am guessing this code is not really getting executed for vllm since we are not selecting specific labels (the labels we select for lora are hardcoded somewhere else). If we have good test coverage it is fine, but again, I would rather have a clear and well defined spec in the model server protocol rather than this.

It's true that most of the string parsing logic isn't being run unless labels are present. There are 2 phases we're planning to go through:
1. Allow any prometheus metric to map to the PodMetrics, i.e. any family with any labels can be KvCacheUtil, WaitingRequestCount, etc. (this requires the ability to filter on arbitrary names/lables)

2. Have a specific metrics family and name dedicated to Gateway that model servers must implement with labels for each PodMetric we need for load balancing. We would no longer need complex parsing, as we'd be looking for a very specific metric format that we could hardcode the scaping for

For this PR (and for initial triton support once they have the compound metrics we need) we're on step 1, but if we can get model serves on board, we could have vLLM, Triton, Jetstream, etc. add a prometheus gauge metrics like:
```
gateway_api_inference_metrics{type="kvCacheUtil"} 
gateway_api_inference_metrics{type="waitingRequestCount"}
... etc ...
```
Then we could do away with this parsing, on our end.

ahg-g commented:
/ok-to-test

ahg-g commented:
Can you rebase pls?

BenjaminBraunDev commented:
Note: While it does make the diff more difficult to read, I moved the metrics out of the vllm folder since it would result in multiple backend/metrics packages, and the code it now no longer vllm specific.

ahg-g commented on pkg/epp/backend/metrics/metrics.go:
```diff
@@ -0,0 +1,272 @@
+	port int32,
+) (*Metrics, error) {
+	logger := log.FromContext(ctx)
+	loggerDefault := logger.V(logutil.DEFAULT)
```
no need to do any logging inside this function, leave it to the caller to do it.

BenjaminBraunDev replied:
removed all logging here, errors are returned up to caller

ahg-g commented on pkg/epp/backend/metrics/metrics.go:
```diff
@@ -0,0 +1,272 @@
+	var latestTimestamp int64 = -1 // Initialize to -1 so any timestamp is greater
+
+	var labels map[string]string = nil
+	if spec.Labels != nil {
+		labels = spec.Labels
+	}
```
why do you need to define this, just use spec.Labels directly

BenjaminBraunDev replied:
will remove, this is an artifact from when I thought spec might be nil. It's redundant.

ahg-g commented on pkg/epp/backend/metrics/metrics.go:
```diff
@@ -0,0 +1,272 @@
+	}
+
+	if latestMetric != nil {
+		logger.V(logutil.TRACE).Info("Labeled metric found", "value", latestMetric, "name", spec.MetricName)
```
ditto, lets remove this log

BenjaminBraunDev replied:
done

ahg-g commented on pkg/epp/backend/metrics/metrics.go:
```diff
@@ -0,0 +1,272 @@
+func (p *PodMetricsClientImpl) getMetric(logger logr.Logger, metricFamilies map[string]*dto.MetricFamily, spec MetricSpec) (*dto.Metric, error) {
+	mf, ok := metricFamilies[spec.MetricName]
+	if !ok {
+		logger.V(logutil.TRACE).Error(nil, "Metric family not found", "name", spec.MetricName)
```
ditto, don't log here pls

BenjaminBraunDev replied:
done

BenjaminBraunDev commented on pkg/epp/backend/metrics/metrics.go:
```diff
@@ -0,0 +1,272 @@
+		}
+	}
+	if latest == nil {
+		logger.V(logutil.TRACE).Info("Metric value Empty", "value", latest, "metric", p.MetricMapping.LoraRequestInfo.MetricName)
```
done

BenjaminBraunDev commented on cmd/epp/main.go:
```diff
@@ -92,6 +91,17 @@ var (
 			"are assumed to be named tls.crt and tls.key, respectively. If not set, and secureServing is enabled, "+
 			"then a self-signed certificate is used.")
+	// metric flags
+	totalQueuedRequestMetric = flag.String("totalQueuedRequestMetric",
```
done

BenjaminBraunDev commented on cmd/epp/main.go:
```diff
@@ -92,6 +91,17 @@ var (
+		"vllm:gpu_cache_usage_perc",
+		"Prometheus metric for the fraction of KV-cache blocks currently in use (from 0 to 1).")
+	// LoRA metrics
+	loraRequestInfoMetric = flag.String("loraRequestInfoMetric",
```
done

BenjaminBraunDev commented on pkg/epp/backend/metrics/metrics.go:
```diff
@@ -0,0 +1,272 @@
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		loggerDefault.Error(err, "Failed create HTTP request", "method", http.MethodGet, "url", url)
```
done

BenjaminBraunDev commented on pkg/epp/backend/metrics/metrics.go:
```diff
@@ -0,0 +1,272 @@
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		loggerDefault.Error(err, "Failed to fetch metrics", "pod", pod.NamespacedName)
```
done

BenjaminBraunDev commented on pkg/epp/backend/metrics/metrics.go:
```diff
@@ -0,0 +1,272 @@
+	}()
+
+	if resp.StatusCode != http.StatusOK {
+		loggerDefault.Error(nil, "Unexpected status code returned", "pod", pod.NamespacedName, "statusCode", resp.StatusCode)
```
done

BenjaminBraunDev commented on pkg/epp/backend/metrics/metrics.go:
```diff
@@ -0,0 +1,272 @@
+
+	loraRequests, ok := metricFamilies[p.MetricMapping.LoraRequestInfo.MetricName]
+	if !ok {
+		logger.V(logutil.TRACE).Error(nil, "Metric family not found", "name", p.MetricMapping.LoraRequestInfo.MetricName)
```
done

BenjaminBraunDev commented on pkg/epp/backend/metrics/metrics.go:
```diff
@@ -0,0 +1,272 @@
+// reason its specially fetched is because each label key value pair permutation generates new series
+// and only most recent is useful. The value of each series is the creation timestamp so we can
+// retrieve the latest by sorting the value.
+func (p *PodMetricsClientImpl) getLatestLoraMetric(logger logr.Logger, metricFamilies map[string]*dto.MetricFamily) (*dto.Metric, time.Time, error) {
```
done

BenjaminBraunDev commented on pkg/epp/backend/metrics/metrics.go:
```diff
@@ -0,0 +1,272 @@
+			}
+		}
+		// Skip if it does not have the lora labels
+		if !hasRequiredLabels {
+			continue
+		}
```
That's true, removed this section and related vars

BenjaminBraunDev commented:
> Pls keep the yaml files where they are for now, the guides reference them.

Moved these files back.

ahg-g commented on pkg/epp/backend/metrics/metrics.go:
```diff
@@ -0,0 +1,249 @@
+	existing *Metrics,
+	port int32,
+) (*Metrics, error) {
+	logger := log.FromContext(ctx)
```
remove this

ahg-g commented on pkg/epp/backend/metrics/metrics.go:
```diff
@@ -0,0 +1,249 @@
+
+// promToPodMetrics updates internal pod metrics with scraped Prometheus metrics.
+func (p *PodMetricsClientImpl) promToPodMetrics(
+	logger logr.Logger,
```
remove this pls

BenjaminBraunDev replied:
done

ahg-g commented:
Thanks for the quick back and forth, this is perfect!

/lgtm
/approve
